src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/MessagePackHubProtocol.java

```diff
@@ -0,0 +1,572 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+package com.microsoft.signalr;
+
+import java.io.IOException;
+import java.math.BigInteger;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.msgpack.core.MessageBufferPacker;
+import org.msgpack.core.MessageFormat;
+import org.msgpack.core.MessagePack;
+import org.msgpack.core.MessagePackException;
+import org.msgpack.core.MessagePacker;
+import org.msgpack.core.MessageUnpacker;
+import org.msgpack.value.ValueType;
+
+class MessagePackHubProtocol implements HubProtocol {
+    
+    private static final int ERROR_RESULT = 1;
+    private static final int VOID_RESULT = 2;
+    private static final int NON_VOID_RESULT = 3;
+
+    @Override
+    public String getName() {
+        return "messagepack";
+    }
+
+    @Override
+    public int getVersion() {
+        return 1;
+    }
+
+    @Override
+    public TransferFormat getTransferFormat() {
+        return TransferFormat.BINARY;
+    }
+
+    @Override
+    public HubMessage[] parseMessages(String payload, InvocationBinder binder) {
+        if (payload.length() == 0) {
+            return new HubMessage[]{};
+        }
+
+        List<HubMessage> hubMessages = new ArrayList<>();
+        
+        try {
+            MessageUnpacker unpacker = MessagePack.newDefaultUnpacker(payload.getBytes(StandardCharsets.ISO_8859_1));
```

Comment:
Looks like we're missing the length prefix for reading and writing messages:
https://github.com/dotnet/aspnetcore/blob/master/src/SignalR/docs/specs/HubProtocol.md#binary-encoding

Reply:
I don't see the C# implementation doing this either: https://github.com/dotnet/aspnetcore/blob/0d156487aa27df65d68fc7fa31e7b6cb9f4d7581/src/SignalR/common/Protocols.MessagePack/src/Protocol/MessagePackHubProtocolWorker.cs#L24-L63. I'm not sure where the requirement for that prefix comes from, I couldn't find it in the messagePack spec

Reply:
It's the `BinaryMessageParser.TryParseMessage` method.
https://github.com/dotnet/aspnetcore/blob/0d156487aa27df65d68fc7fa31e7b6cb9f4d7581/src/SignalR/common/Shared/BinaryMessageParser.cs#L13
And described at the end here: https://github.com/dotnet/aspnetcore/blob/master/src/SignalR/docs/specs/HubProtocol.md#binary-encoding

Reply:
https://github.com/dotnet/aspnetcore/pull/23532/commits/8ab171aed3ce4aa8017b4b50f5590917fe356170

Reply:
38c37a9548678b7a4c13d78b78f488d72f9dcf46

Reply:
Need to fix this so we don't use packer/unpacker for header